### ISSUE TITLE
Suggestion for minor option addition to the typescript-eslint/member-delimiter-style rule

### ISSUE THREAD

**Issue by doberkofler** [typescript-eslint/typescript-eslint#162]
**Repro**

The `typescript-eslint/member-delimiter-style` rule has already several option but at least for my coding guidelines a minor enhancement would go a long way.
The `multiline.requireLast` flag allows to either force or forbid using a delimiter in the last line and it would be great to add an option that allows both variants.

```JSON5
{
  "rules": {
    '@typescript-eslint/member-delimiter-style': ['warn',
	{
		singleline: {
			delimiter: 'comma',
			requireLast: false
		},
		multiline: {
			delimiter: 'comma',
			requireLast: false
		}
	}
    ]
  }
}
```

```TS
type foo = {
   foo: string,
   bar: string,
};

type bar = {
   foo: string,
   bar: string
};
```

**Expected Result**

For example allowing to omit the `requireLast` flag or set to to `null` indicating that a delimiter is allows but not mandatory.

**Actual Result**

In the above example `bar: string,` fails because the delimiter is given.

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `1.1.1` |
| `@typescript-eslint/parser`        | `1.1.0` |
| `TypeScript`                       | `3.2.4` |
| `ESLint`                           | `5.12.1` |
| `node`                             | `10.15.0` |
| `npm`                              | `6.7.0` |

**Comment by bradzacher** [comment]
I'm not sure if we want to add an option to disable the require check.
ESLint rules are about enforcing consistency in your codebase.

If we allow you to not set a value for requireLast, then we're allowing you to set the rule to **not** enforce consistency in your codebase.

Other members / community can weigh in, but my gut feel is that this is not a good option.

As an example, this is a similar core eslint rule `comma-dangle`, wihch only has the option to enforce always or never: https://eslint.org/docs/rules/comma-dangle

**Comment by doberkofler** [comment]
I would find it very helpful but understand your point

**Comment by unional** [comment]
@bradleyayers I have the same need as @doberkofler .

The `comma-dangle` rule has a `only-multiline` option which ignores the last comma.

**Comment by bradzacher** [comment]
I'm not sure I understand why it matters @unional.
Types have nothing to do with the javascript version you're compiling to - they are not emitted.

So why does it matter if you have a trailing comma or not?

**Comment by unional** [comment]
@bradzacher Yes, you are right about that.
The reason I want to add that in so that the TypeScript and JavaScript coding will be the same.

i.e.:
```
type Foo = {
  a: number,
  b: string
}

const foo: Foo = {
  a: 1,
  b: 'abc'
}
```

**Comment by bradzacher** [comment]
QQ - why are you targetting ES3 code?
What's the use case for shipping unbundled, unminified ES3 compatible code?

If you do have a bundle/minification step - then why are you bothering with writing es3 compatible syntax?

**Comment by doberkofler** [comment]
My motivation was about following my existing coding guidelines in a very large repository and I now had to disable the rule and allow inconsistent use of the delimiter.

My rule would be:
1) Force the code to always use comma as delimiter is the important part.
2) Force the code to omit the delimiter on single line definitions but allow it on multi line.

I still believe this would be a nice addition to this rule and now it seems I’m not the only one.

**Comment by unional** [comment]
I have similar argument as @doberkofler.

The style I maintain in [`eslint-plugin-harmony`](https://github.com/unional/eslint-plugin-harmony) it target for ease of use while maintaining consistency.

The tailing comma in multiline is "nice to have" but not an absolute necessity.
For example, the style prefer ASI as in `standard` but opt for `semi-standard` because there are cases (although VERY rarely) that semi-colon is needed, but I would not enforce the rule so that the only way to handle that case is like this:

```ts
const x = foo()

;[1,2,3].forEach(...)
```

because that will screw up indentation in many IDE including VSCode.

**Comment by bradzacher** [comment]
I can partially understand asking for a way to enforce never adding the final trailing delimiter. (I say partially because on the one hand it would give it parity with comma-damgle, but on the other hand there's no reason for it because types/interfaces aren't held down by the same es3 compatibility)

But I do not understand why you'd ask for an option to ignore whether or not it exists.

There's a fixer, and eslint has IDE integrations which puts the fix one click away.

Why would you allow your codebase to be inconsistent when there is actually zero cost to make and enforce it stays consistent?

How come you don't just turn the rule on, run the fixer and then commit it?

**Comment by unional** [comment]
Believe or not, there are teams that do not use any auto-fixer.

Again, it is also about breaking the flow. If this is enforced, then while the developer is typing the code, these errors will keep popping up until he save the file and let the auto-fixer kicks in.

So while he is work on his craft, these errors become a distraction to him and he have to keep wondering "is that an actual error or it will go away when I save the file".

This slows down the developer and is not a present experience.

**Comment by bradzacher** [comment]
> Believe or not, there are teams that do not use any auto-fixer.

I understand that people don't turn on autofix on save. I don't have that turned on because I don't always want things to be autofixed because sometimes I violate lint rules intentionally whilst iterating.

In regards to the fixer, I'm talking about two things:

First and foremost - onboarding your engineers onto the rule, and what it enforces. The fixer only needs to be used once for them to learn how to fix the error (if the error message wasn't enough for them).

Second, the fixer works well for _implementing_ the rule on your codebase. You don't have to worry about turning it on and having to fix the violations manually, or turning it on and having engineers run into this rule on old code.

If you're worried about accidentally running fixers from other rules, it's pretty simple to just run the one fixer from the CLI:

`eslint --plugin @typescript-eslint --parser @typescript-eslint/parser --rule '@typescript-eslint/member-delimiter-style: [2, {config...}]' <files here>`
https://eslint.org/docs/user-guide/command-line-interface

> So while he is work on his craft, these errors become distractions to him and he has to keep wondering "is that an actual error or it will go away when I save the file".
>
> This slows down the developer and is not a present experience.

I don't think I've ever met a developer that will repeatedly, and continually see the same error in the same place, and not learn what it means or how to fix it.

I've worked with people with 20 years experience, and people with 6 months. They all were able to see lint errors, recognise the fix and learn to stop it happening.

**Comment by doberkofler** [comment]
@bradzacher I completely understand why you fight not to overload the number of options but at the end this come down to taste. I would really like to keep our style but also to enforce it.
We used fix many times where we had to fix inconsistencies or decided to adopt a change in style but we really like the current style and would like to enforce it not to change it.
Our style is to enforce what delimiter to use but relax when it must be used and I actually have see this quite in bit in other code bases as well.

**Comment by unional** [comment]
@doberkofler I have modified the rule and published it under [`eslint-plugin-harmony`](https://github.com/unional/eslint-plugin-harmony/blob/master/docs/rules/member-delimiter-style.md) if you are interested.

**Comment by doberkofler** [comment]
@unional Looks good. I think you should open a pull request here.
